src/js/init.s

```diff
@@ -118,6 +122,51 @@ SORT_START_ROW    = 3
 SORT_START_ROW    = 2
 .endif
 
+; .ifdef _CROWD_CONTROL_
+
+GAME_MODE_CHANGE_LOCATION = $01
+OP_JMP_ABS  = $4c
+
+; Short How-To-Use documentation:
+;
+; 0. Make sure that there isn't already anything waiting to be run by seeing if
+; CrowdControlFlag is zero. This gets cleared after the effects are completed.
+;
+; 1. Set one or more flags in the CrowdControlQueue field to run that operation
+; the next time the game is available.
+;
+; 2. If you need to update the status bar for an operation, set the appropriate bit
+; in CrowdControlStatusBarUpdate.
+;
+; 3. Set CrowdControlFlag to 1, and the game will run the update the next time the
+; player gets back to "normal" play state.
+;
+
+; If set, during the next main loop, run the next CC op from the queue
+; INTERNAL - bit 7 will be set when we need to early return (after wild warp)
+CrowdControlFlag              = $6220
+
+; Bit packed operations that Crowd Control can update.
+; %000000ws
+;        |+-- s = status bar update. See CrowdControlStatusBarUpdate
+;        +--- w = Wild Warp. Uses the next wild warp location for the user.
+;                 ?? Should we just always go back to mezzaine to prevent softlocks?
+CrowdControlQueue             = $6221
+
+; Bit packed list of status bar fields to update
```

Comment:
This is sort of exactly what I was worrying about w.r.t. version skew.
As much as possible, I'd rather the ObjC side of the CC script just be a "dumb shell" with all the smarts/logic here.

My thought is that we have a ton of free space in the expanded ROM pages that we could use for this, so we could actually be relatively verbose with the code to handle it all.  Then we embed a table in the ROM for the ObjC to read, which would tell it all the CC capabilities.  Potentially this could be as extensive as - for each effect, its name, cost, and an address and value that needs to be written to accomplish it.  Then we could have one or more addresses that we'd check at various times to do things.  For instance, 6220 might be for the main control loop right after reading controller input, while 6221 could be after processing all the object actions (for instance).  Maybe there's not really any other timings we'd need to worry about, but the table makes it easy to only check for what we're ready to handle.  Then the main game code checks for nonzero (or non-FF, whatever) and if it sees something it pages in the extended page and indexes a jump table.

In this setup, the game itself would know what needed to be updated.  And since all the logic is in the ROM, coming from the randomizer, there's no worry about version skew - the ObjC script is trivial and never gets out of date.  And you've got a single source of truth if the user wants to change the options or prices, or disable certain effects, etc.

Reply:
I'm not against doing this, but I'm also not the one that wants to do all those changes. I only wrote this to help out @TheSirArchibald as it stands, and he's the one working on the C# code for it. So that leaves the problem of who's going to do the ASM work for it.

Reply:
That's definitely a fair point :-)